datafusion/common/src/hash_utils.rs

```diff
@@ -484,6 +485,40 @@ fn hash_fixed_list_array(
     Ok(())
 }
 
+#[cfg(not(feature = "force_hash_collisions"))]
+fn hash_run_array<R: RunEndIndexType>(
+    array: &RunArray<R>,
+    random_state: &RandomState,
+    hashes_buffer: &mut [u64],
+    rehash: bool,
+) -> Result<()> {
+    let values = array.values();
+    let values_len = values.len();
+    let mut values_hashes = vec![0u64; values_len];
+    create_hashes(&[Arc::clone(values)], random_state, &mut values_hashes)?;
+
+    let run_ends = array.run_ends();
+    let mut prev_run_end = 0;
+
+    for (i, value_hash) in values_hashes.iter().enumerate().take(values_len) {
+        let run_end = run_ends.values()[i].as_usize();
+
+        if rehash {
+            for hash in hashes_buffer.iter_mut().take(run_end).skip(prev_run_end) {
+                *hash = combine_hashes(*value_hash, *hash);
+            }
+        } else {
+            for hash in hashes_buffer.iter_mut().take(run_end).skip(prev_run_end) {
```

Comment:
This could be optimized to `hashes_buffer[prev_run_end..run_end].fill(value_hash)` (SIMD friendly!).